linera-core/src/client/mod.rs

```diff
@@ -1395,11 +1395,63 @@ where
                 return Err(error.into());
             }
         };
-        for received_certificates in responses {
-            // Process received certificates.
-            self.receive_certificates_from_validator(received_certificates)
-                .await;
+        let response_count = responses.len();
+        let mut other_sender_chains = BTreeSet::new();
+        let mut certificates = BTreeMap::<ChainId, BTreeMap<BlockHeight, Certificate>>::new();
+        for response in responses {
+            other_sender_chains.extend(response.other_sender_chains);
+            // Update the tracker: Even if the validator sent us unsuitable certificates, there is
+            // no need to ask them again.
+            self.state_mut()
+                .update_received_certificate_tracker(response.name, response.tracker);
```

Comment:
I `receive_certificates_from_validator` this method is called at the very end - after `receive_certificate_internal` and `retry_pending_cross_chain_requests` . Is that not a problem that you changed the order?

Reply:
That's why I added the comment above: In my opinion it's fine to update the tracker here, but I'm happy to reconsider.

Reply:
I didn't catch that the comment is explaining why we can do it here, instead of at the end.

Reply:
On second thought, maybe it's better to set all trackers at the end, you're right.

My thought was that since we already checked the signatures, processing the certificates should always succeed anyway.

But of course if we have an issue locally, or the process just gets killed before processing everything, then we _shouldn't_ update the trackers.

Reply:
No, you're right, I'll move them to the end.